Fix compile errors

leg0 opened the pull request:
Fix compile errors in Wire and UART.
 - added namespace qualifications where necessary
 - fixed conflicting return types in overrides
 - removed requestFrom overloads that don't add value and make the function more difficult to use.

awatterott committed:
No fixed value for USB power current.

facchinm committed:
[ArduinoAPI] initial porting

giampiero7 committed:
Added SERIAL_PORT_MONITOR define to Arduino Ethernet

Port of https://github.com/arduino/Arduino/pull/7023

greyltc committed:
Introduce non compulsory Wire timeout

move timout handling into its own function

change timeout from milliseconds to microseconds

don't forget operating slave address or the bitrate when we reset because of a timeout

Co-Authored-By: Witold Markowski <witold.a.markowski@gmail.com>

fix delay datatype uint16_t --> uint32_t

Update libraries/Wire/src/utility/twi.c

fix mix up using TWBR instea of TWAR!

Co-Authored-By: Matthijs Kooijman <matthijs@stdin.nl>

Update libraries/Wire/src/utility/twi.c

fix 2nd TWBR/TWAR mixup

Co-Authored-By: Matthijs Kooijman <matthijs@stdin.nl>

twi_stop() should use the same timeout as everywhere else

all while loops are now protected by timeouts

Revert "twi_stop() should use the same timeout as everywhere else"

This reverts commit 68fe5f1dae1bb41183bb37eeda3fb453394a580c.

make timeout counter volatile

rename timeout function for improved clarity

- resetting the twi interface on timeouts is now optional
- timeouts in the ISR are no longer hardcoded and now obey the set timeout value
- a user-readable flag is now set whenever a timeout occurs
  - the user can clear this flag whenever they like

jamadagni committed:
Add `bitToggle` macro to complement `bitSet` etc

Pharap committed:
Add placement new operator

Pharap committed:
Prevent possible compiler warning

Casting to void is a well known trick for prevening 'unused parameter' warnings.

Pharap committed:
Add noexcept specifier to placement new operator

The standard mandates that placement new should be have a noexcept specifier.

florian-schweiger committed:
Removed historic Arduino-0012 workaround

Removed #undefs in SoftwareSerial.h that were marked as Arduino 0012 workaround and that broke several macros, including abs.
See https://github.com/arduino/ArduinoCore-avr/issues/30

algernon committed:
Add USBDevice.isSuspended()

Based on code originally by Rob van der Veer <rob.c.veer@gmail.com>, this adds
USBDevice.isSuspended(), so user sketches can run custom code in their `loop`
methods after checking if the device is suspended or not.

Signed-off-by: Gergely Nagy <algernon@keyboard.io>

per1234 committed:
Revert "Updated Arduino as ISP setting"

This reverts commit b084848f2eaf9ccb3ac9a64ac5492d91df4706bf.

per1234 committed:
Add ATmega32U4-compatible Arduino as ISP programmer

Use of the stk500v1 protocol for Arduino as ISP does not work with native USB boards on Windows. The arduino protocol does.

However, the arduino protocol makes it more likely that boards with an external USB interface chip will require the auto-reset circuitry to be disabled to allow them to be used as Arduino as ISP. That adds extra complexity to a process already difficult for the average Arduino user.

For this reason, a new programmer using the arduino protocol is added specifically for using native USB boards as Arduino as ISP and the previous Arduino as ISP configuration is retained for use with all other boards.

facchinm committed:
Merge pull request #56 from per1234/arduino-as-isp-atmega32u4-programmer

Add ATmega32U4-compatible Arduino as ISP programmer

facchinm committed:
Merge pull request #54 from algernon/pull/wakeup-suspend-hooks

Add USBDevice.isSuspended()

paulo-raca committed:
Remove commented out code for I2C interrupts on WInterrupts.c

These are currently implemented by the Wire library, on twi.c

facchinm committed:
Merge pull request #59 from paulo-raca/wiring_private_i2c

Remove commented out code for I2C interrupts on WInterrupts.c

paulrnash committed:
Add parameters names to common prototypes

Putting the parameter names in these commonly used prototypes makes syntax help like VS Code's Intellisense work 99% more useful. Without them, it doesn't give you the names of the parameters and you have to remember the semantics yourself. :(

dmadison committed:
Define descriptor return var in pluggable block

Avoids unused variable warning if USB is enabled but pluggable USB is not

luco5826 committed:
Minor optimization in shiftOut function

luco5826 committed:
Fixed shiftOut in MSBFIRST mode HIGH/LOW

facchinm committed:
Merge pull request #51 from florian-schweiger/patch-1

Remove historic Arduino 0012 workaround in SoftwareSerial

matthijskooijman committed:
Do not claim AT-protocol in CDC interface descriptor

The CDC code presents itself as a virtual serial port. However, it also
sets the "bFunctionProtocol" value to 1, which means it supports
AT-commands, which is not actually the case. This might cause problems
with some software, such as ModemManager.

Originally, ModemManager would be very liberal with probing serial
devices, using a blacklist to prevent probing non-modems such as
Arduinos.

Since version 1.7.990, it has supported a "strict" mode where it tries to be
more restrained in what devices it probes. For CDC ACM devices, this
means it will only probe devices that claim to support AT-commands.
However, it also stopped applying the blacklist (intending to eventually
remove the blacklist), meaning it would again probe Arduinos.

This new strict policy is not the upstream default, but is enabled in
Debian (since Buster) and Ubuntu (since bionic 18.04.2).

The proper way to fix this, is to not claim AT comand support in the USB
device descriptor, which is what this commit does. The Arduino will
still show up as a virtual serial port, just not be probed by
ModemManager in strict mode.

For the commit that introduced the strict mode in ModemManager, see
https://cgit.freedesktop.org/ModemManager/ModemManager/commit/src?id=ee570d44dc117dc69f23e83313dd877f76c5e3e0

jrowberg committed:
Cast pins to signed integers to avoid Wtype-limits compile warning

jrowberg committed:
Add null pointer test to String destructor

gigoh committed:
rename yunmini.bootloader.file on boards.txt

CombiesGit committed:
Update twi.c

Added  __attribute__ ((fallthrough));

zakkie committed:
remove unnecessary if branch

checking `length` in below while statement

aentinger committed:
Adding parenthesis around 'bitvalue' allowing correct macro expansion when using with ternary operator such as bitWrite(value, bit, some_computed_value == 5 ? 1: 0);'

aentinger committed:
Merge pull request #103 from zakkie/fix/remove_if

Removes unnecessary if branch (because length is checked in while statement below the if-clause).

aentinger committed:
Merge pull request #76 from dmadison/pluggable-unused

Fix unused variable warning for non-pluggable USB in SendDescriptor

aentinger committed:
Merge pull request #64 from paulrnash/paulrnash-intellisense

Add parameter names to common prototypes

aentinger committed:
Merge pull request #33 from Pharap/add-placement-new

Add placement new operator

aentinger committed:
Merge branch 'master' into patch-1

aentinger committed:
Merge pull request #31 from jamadagni/patch-1

Add `bitToggle` macro to complement `bitSet` etc

aentinger committed:
Merge pull request #100 from CombiesGit/patch-1

[TWI] Add __attribute__ ((fallthrough)) to remove recent GCC warnings

aentinger committed:
Merge pull request #99 from gigoh/fix-bootloader-path

Correct yunmini.bootloader.file on boards.txt

aentinger committed:
Merge pull request #92 from matthijskooijman/cdc-no-prototocol-at

Do not claim AT-protocol in CDC interface descriptor

aentinger committed:
Merge pull request #95 from jrowberg/master

Cast pins to signed integers to avoid Wtype-limits compile warning

aentinger committed:
Merge pull request #82 from luco5826/master

Minor optimization in shiftOut function

Corjan85 committed:
Changed linking order, so precompiled libraries can use the Arduino c… (#52)

* Changed linking order, so precompiled libraries can use the Arduino code functions.
* Added 'compiler.libraries.ldflags', reverted object order in previous commit

asukiaaa committed:
Add i2c_scanner to example of Wire (#93)

aentinger committed:
Merge pull request #4 from awatterott/master

No fixed value for USB power current.

aentinger committed:
Revert "Changed linking order, so precompiled libraries can use the Arduino c… (#52)"

This reverts commit 41f15a1359943ccb3ec9666a2e28a2d15e9581fc.

aentinger committed:
Merge pull request #284 from arduino/revert-52-master

Revert "Changed linking order, so precompiled libraries can be used"

mastrolinux committed:
Remove Genuino Occurrences

cmaglie committed:
Removed Genuino Micro and set to replace it with Arduino Micro on upload

facchinm committed:
Merge pull request #286 from arduino/remove-genuino

Remove Genuino Occurrences

M-Reimer committed:
Interrupt ordering for 32u2 and 16u2 MCU (#66)

* Interrupt ordering for 32u2 and 16u2 MCU
* Added missing chip variants

cmaglie committed:
Release 1.8.2

e28eta committed:
Fix sine -> since typo in HardwareSerial files

facchinm committed:
Merge pull request #318 from e28eta/typos

Fix sine -> since typo in HardwareSerial files

cmaglie committed:
Added precompiled lib support

Max-Leopold committed:
Update comment on wiring_analog.c

facchinm committed:
Merge pull request #337 from Max-Leopold/patch-1

Correct comment on wiring_analog.c

eudoxos committed:
Add "new" proxy header for compatibility with c++-standard #include<new>

VitSh7 committed:
Improve reading ADC result

7.3.0-atmel3.6.1-arduino7 gcc fails to optimize separate reading from ADCL and ADCH. It produces additionally three eor commands or in some cases two mov commands in the assembly code (see discussion #344). These commands swap register contents before store them to data area. So they are completely unnecessary. Reading ADC result with ADC macro fixes it and gcc generates the right code..

facchinm committed:
Merge pull request #335 from cmaglie/add-precomp-lib-support

Added precompiled library support

facchinm committed:
Merge pull request #340 from eudoxos/header-new

Add "new" proxy header for compatibility with c++-standard #include<new>

facchinm committed:
Wire: add sensible defaults to setWireTimeout

matthijskooijman committed:
Wire: improve comments on timeout

facchinm committed:
Publish 1.8.3

facchinm committed:
Wire: apply last suggested comment from @matthijskooijman

matthijskooijman committed:
Clean up and complete `<new>` header

This makes this header complete up to including C++14, except two
exception classes that cannot be defined without `<exception>`.

The functions related to the "new_handler" are declared but not actually
defined, to prevent overhead and complexity. They are still declared to
allow implementing them in user code if needed.

This makes the implementation of all operator new and delete functions
comply with the C++11/C++14 specification in terms of which should be
actually implemented and which should be delegate to other functions.

There are still some areas where these implementations are not entirely
standards-compliant, which will be fixed in subsequent commits.

This fixes part of #287 and fixes #47.

matthijskooijman committed:
Allow overriding selected operator new and delete functions

This makes these functions weak, so that a sketch or library can replace
them. This does not apply to all of these operators, only for the ones
that the C++ standard specifies as replaceable.

matthijskooijman committed:
Add weak `std::terminate()` implementation

This allows calling it from other places later. The default
implementation calls `abort()`, but making it weak allows user code to
override this function (either directly, or by including a library like
uclibc++ that implements `std::set_terminate()`).

Note that this does not add a declaration for this function, since the
standard dictates this to be in `<exception>`, but we cannot
meaningfully or completely implement that header, so better leave it to
be overridden by e.g. libraries like uclibc++.

matthijskooijman committed:
Call std::terminate on pure or deleted virtual functions

These are special functions that are presumably put into vtables for
deleted or pure virtual functions. Previously, this would call `abort()`
directly, but calling `std::terminate()` achieves the same effect, but
allows user code to change the behavior (e.g. to print to serial, blink
leds or whatever makes sense).

matthijskooijman committed:
Make zero-sized new standards-compliant

This fixes part of #287.

matthijskooijman committed:
Optionally let new terminate on allocation failure

This is currently disabled, keeping the old behavior of returning
NULL on failure, but should probably be enabled in the future as code
that does want to do a null check has had a chance to switch to the
more portable nothrow versions.

When enabled, allocation failure calls the weak `std::terminate()`,
which calls `abort()` by default, but can be replaced by user code to do
more specific handling.

To enable this, a macro must be defined (in new.cpp or on the compiler
commandline).

This fixes part of #287.

matthijskooijman committed:
Swap new and new.h header files

Originally, the Arduino core used "new.h", rather than the standard
"new", probably because the implementation was incomplete, and for the
most commonly used new and delete operators, no include is needed at all
(they are defined implicitly by the compiler). However, now Arduino
does expose the "new" name, as an alias for the older "new.h". Given
that the standard name is "new", it makes more sense to put the actual
content in "new", and make "new.h" a compatibility header that includes
"new" instead of the other way around.

matthijskooijman committed:
Use std::size_t in new/delete

The standard dictates that `std::size_t` is used, rather than the plain
`size_t` type.

Even though these types are usually, if not always, exactly the same
type, other code might assume that `std::size_t` is actually used and thus
also available under that name after including `<new>`.

This fixes that by using the right type. One challenge is that it is
usually declared in headers that we do not have available, so this just
defines the `std::size_t` type in the `<new>` header to work around
that.

facchinm committed:
Merge pull request #361 from matthijskooijman/improve-new

Improve and complete implementation of new/delete

facchinm committed:
Merge pull request #345 from Vitve4/pr_344

Improve reading ADC result

per1234 committed:
Update platform specification URLs in configuration files

The Google Code URL in boards.txt is very outdated. The URL in platform.txt is to a more recent home of the content,
which has been replaced with a link to the new location, but while I'm updating boards.txt, I might as well point both
URLs to the real page.

facchinm committed:
Merge pull request #369 from per1234/update-platform-docs-urls

Update platform specification URLs in configuration files

DanielGibson committed:
Allow disabling CDC with -DCDC_DISABLED

Sometimes Arduino-based USB devices don't work because some hardware
(like KVM switches) gets confused by the CDC sub-devices.
This change makes it relatively easy to disable CDC at compiletime.
Disabling it of course means that the serial console won't work anymore,
so you need to use the reset button when flashing.

CDC_DISABLED is also used in ArduinoCore-samd for the same purpose.

based on
https://github.com/gdsports/usb-metamorph/tree/master/USBSerPassThruLine

See also https://github.com/NicoHood/HID/issues/225 and
https://github.com/arduino/Arduino/issues/6387 and
https://forum.arduino.cc/index.php?topic=545288.msg3717028#msg3717028

greyltc committed:
fix twi_manageTimeoutFlag function description

facchinm committed:
Merge pull request #385 from greyltc/patch-1

fix twi_manageTimeoutFlag function description

ArkadyGamza committed:
Declare TwoWire functions as virtual

To make alternative implementations of the TwoWire class (e.g. SoftwareWire for software I2C) work properly being passed to libraries that expect TwoWire type.

per1234 committed:
Use established KEYWORD_TOKENTYPE for Wire keyword

All the other platform bundled libraries use KEYWORD1 for the library's global instance.

leg0 committed:
Fix compile errors

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/arduino/ArduinoCore-avr?pullRequest=410) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/arduino/ArduinoCore-avr?pullRequest=410) before we can accept your contribution.<br/>**2** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: facchinm<br/>:white_check_mark: per1234<br/>:x: leg0<br/>:x: umbynos<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/arduino/ArduinoCore-avr?pullRequest=410) it.</sub>

aentinger committed:
Merge pull request #396 from ArkadyGamza/make_TwoWire_functions_virtual

Declare TwoWire functions as virtual

per1234 committed:
Provide a brief description of the project

per1234 committed:
Configure Dependabot to check for outdated actions used in workflows

Dependabot will periodically check the versions of all actions used in the repository's workflows. If any are found to
be outdated, it will submit a pull request to update them.
NOTE: Dependabot's PRs will sometimes try to pin to the patch version of the action (e.g., updating `uses: foo/bar@v1`
to `uses: foo/bar@v2.3.4`). When the action author has provided a major version ref, use that instead
(e.g., `uses: foo/bar@v2`). Dependabot will automatically close its PR once the workflow has been updated.
More information:
https://docs.github.com/en/github/administering-a-repository/keeping-your-actions-up-to-date-with-dependabot

per1234 committed:
Add CI workflow to check for commonly misspelled words

On every push, pull request, and periodically, use the codespell-project/actions-codespell action to check for commonly
misspelled words.

In the event of a false positive, the problematic word should be added, in all lowercase, to the ignore-words-list field
of ./.codespellrc. Regardless of the case of the word in the false positive, it must be in all lowercase in the ignore
list. The ignore list is comma-separated with no spaces.

aentinger committed:
Revert "Declare TwoWire functions as virtual" (#412)

per1234 committed:
Correct typos in comments and documentation

per1234 committed:
Add CI workflow to do Arduino project-specific linting

On every push, pull request, and periodically, run Arduino Lint to check for common problems not related to the project
code.

per1234 committed:
Add "smoke test" examples compilation CI workflow

On every push or pull request that affects platform source code or bundled libraries, compile example
sketches of all bundled libraries for the platform's boards.

per1234 committed:
Report changes in memory usage that would result from merging a PR

On creation or commit to a pull request, a report of the resulting change in memory usage of the examples will be
commented to the PR thread.

facchinm committed:
Merge pull request #413 from per1234/ci

Use GitHub Actions for continuous integration

per1234 committed:
Don't use codespell's "rare" dictionary for CI spell check

The word "statics" was recently added to codespell's "rare" commonly mispelled words directory. By default, this
dictionary is used in addition to the more conservative "clear" dictionary, resulting in a spurious failure of the
"Spell Check" CI workflow.

Although this could be fixed by adding the word to the ignore list, it seems that an increasing number of common words
being added to this dictionary, so I think it's best to just remove it altogether. The goal for the "Spell Check"
workflow is to catch some typos without a significant number of false positives, even if that means some misspellings
slip through. I think the "clear" dictionary is more in line with that goal.

silvanocerza committed:
Merge pull request #417 from per1234/ignore-word

Don't use codespell's "rare" dictionary for CI spell check

facchinm committed:
Merge pull request #383 from DanielGibson/disable-cdc

Allow disabling CDC with -DCDC_DISABLED

cmaglie committed:
Added pluggable discovery/monitor definitions

felias-fogg committed:
Inserted cast to unsigned int in available method in order to avoid call to __divmodhi4

facchinm committed:
Merge pull request #433 from felias-fogg/master

Speed up 'available' method by inserting unsigned int cast

facchinm committed:
Merge pull request #426 from arduino/scerza/pluggable-discovery-support

Add support for pluggable discovery

facchinm committed:
Add Uno mini board

facchinm committed:
Release 1.8.4

silvanocerza committed:
Fix legacy upload.network_pattern rule

silvanocerza committed:
Merge pull request #440 from arduino/scerza/fix-java-ide-network-upload

Fix legacy `upload.network_pattern` rule

umbynos committed:
initial import from https://github.com/arduino/Arduino/tree/master/build/windows/dist/drivers (moved also gemma stuff under `gemma/`)

umbynos committed:
modify post_install.bat according to https://github.com/arduino/ArduinoCore-megaavr/blob/master/post_install.bat

umbynos committed:
add install for arduino gemma drivers: they are in another directory

umbynos committed:
add drivers for Uno Mini

umbynos committed:
Merge pull request #442 from umbynos/umbynos/add_drivers_install

Add drivers originally bundled with Java IDE directly in the core

per1234 committed:
Add unomini to sketch compilation CI workflow

The "Compile Examples" GitHub Actions workflow provides a basic "smoke test" for the platform by compiling the relevant example sketches for each of the boards on every commit and pull request.

A new board has been added to the platform (Uno Mini) and so should be added to the list of boards compiled for in this workflow.

Even though the Uno compilation does provide reasonable coverage for the Uno Mini due to their similar configurations, that coverage does not extend to the board definition itself.

per1234 committed:
Use more logical Boards menu placement for Uno Mini

Arduino has added a new board to the AVR-based line: the Uno Mini. The Uno Mini has its own board definition in this
platform, which produces a dedicated item in the **Tools > Board** menu of the Arduino IDE.

The classic Arduino IDE arranges the Boards menu for each platform according to the order of occurrence of the board
definition in the platform's boards.txt configuration file. The previous placement of the Uno Mini board definition
resulted in the board appearing at the end of the menu, along with retired and 3rd party boards.

The new placement will cause it to be shown alongside the related Arduino Uno and in a position befitting the latest
official AVR board.

per1234 committed:
Merge pull request #448 from per1234/add-unomini-ci

Add unomini to sketch compilation CI workflow

facchinm committed:
Merge pull request #449 from per1234/unomini-menu-location

Use more logical Boards menu placement for Uno Mini

facchinm committed:
Release 1.8.5

dependabot[bot] committed:
Bump actions/checkout from 2 to 3

Bumps [actions/checkout](https://github.com/actions/checkout) from 2 to 3.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/v2...v3)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

per1234 committed:
Merge pull request #465 from arduino/dependabot/github_actions/actions/checkout-3

Bump actions/checkout from 2 to 3

dependabot[bot] committed:
Bump actions/upload-artifact from 2 to 3

Bumps [actions/upload-artifact](https://github.com/actions/upload-artifact) from 2 to 3.
- [Release notes](https://github.com/actions/upload-artifact/releases)
- [Commits](https://github.com/actions/upload-artifact/compare/v2...v3)

---
updated-dependencies:
- dependency-name: actions/upload-artifact
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

per1234 committed:
Merge pull request #470 from arduino/dependabot/github_actions/actions/upload-artifact-3

Bump actions/upload-artifact from 2 to 3

per1234 committed:
Merge pull request #404 from per1234/patch-1

Use established KEYWORD_TOKENTYPE for Wire keyword
